### PR TITLE
Remove guard_size_oblivious from default contiguity python check, and add aten.sym_is_contiguous.

### PR DESCRIPTION
**Pull request by laithsakka**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #159197

This might cause some new DDEs on call sites that do not use is_contiguous_or_false() or sym_is_contiguous()
but want to find those call sites to handle this properly by calling  is_contiguous_or_false() and not is_contiguous() explitly when appropriate. 
I had to fix one issue after removing the implicit size oblivious reasoning. here is context

we defined in this https://github.com/pytorch/pytorch/pull/157472 sym_is_contiguous to be the function computing contiguity for dynamic shapes in c++. It returns a symbolic expression that represents contiguity and guaranteed not to throw a DDE.

when people call is_contiguous we do sym_is_contiguous().guard_bool()
when people call is_contiguous_or_false we do sym_is_contiguous().guard_or_false()

one issue not handled well was this path
```
c10::SymBool TensorImpl::sym_is_contiguous_custom(
    at::MemoryFormat memory_format) const {
  if (C10_UNLIKELY(matches_python_custom(SizesStridesPolicy::CustomStrides))) {
    return pyobj_slot_.load_pyobj_interpreter()->is_contiguous(
        this, memory_format);
  }

  return sym_is_contiguous_default(memory_format);
}
```
namely if we call sym_is_contiguous_custom but we have matches_python_custom(SizesStridesPolicy::CustomStrides) return true , then we used to call is_contiguous(this, memory_format);

This used to go through the load_pyobj_interpreter and end up calling the python is_contiguous call which used implicit size oblivious reasoning.
once we removed that implicit size oblivious reasoning, the right thing we want is to call
return pyobj_slot_.load_pyobj_interpreter()->sym_is_contiguous(this, memory_format);
otherwise we would get DDE even if the caller is doing sym_is_contiguous.

so I had to define it for pyinterpreter, and then I had to override it for nested tensors.

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @chenyang78 @kadeng @chauhang @amjames @Lucaskabela

### CODE REVIEW

File: torch/_prims_common/__init__.py [hunk 1]
```diff
@@ -265,12 +265,14 @@ def is_contiguous(a: TensorLikeType, false_if_dde=False) -> bool:
     from torch.fx.experimental.symbolic_shapes import (
         guard_or_false,
         guard_or_true,
-        guard_size_oblivious,
         is_nested_int,
     )
 
-    maybe_guard_or_false = guard_or_false if false_if_dde else guard_size_oblivious
-    maybe_guard_or_true = guard_or_true if false_if_dde else guard_size_oblivious
+    def eval_eager(x):
```
Comment (laithsakka) [inline torch/_prims_common/__init__.py:271]: shall we just do 
```
if false_if_dde:
   return torch.ops.aten.sym_is_contiguous.default(input).guard_or_false here?
```
make debugging harder, but make impl simpler.

File: c10/core/TensorImpl.cpp [hunk 2]
```diff
@@ -313,7 +313,7 @@ void TensorImpl::throw_data_ptr_access_error() const {
 c10::SymBool TensorImpl::sym_is_contiguous_custom(
     at::MemoryFormat memory_format) const {
   if (C10_UNLIKELY(matches_python_custom(SizesStridesPolicy::CustomStrides))) {
-    return pyobj_slot_.load_pyobj_interpreter()->is_contiguous(
```
Comment (laithsakka) [inline c10/core/TensorImpl.cpp:316]: calling this causes DDE. we need to delegate to sym_is_contiguous and define it in python for those that overrides it

File: aten/src/ATen/native/native_functions.yaml [hunk 3]
```diff
@@ -5496,6 +5496,13 @@
   tags: core
   manual_cpp_binding: True
 
+- func: sym_is_contiguous(Tensor self, MemoryFormat memory_format=contiguous_format) -> SymBool
+  variants: function
+  device_check: NoCheck
+  device_guard: False
+  tags: core
+  manual_cpp_binding: True
```
Comment (zou3519) [inline aten/src/ATen/native/native_functions.yaml:5517]: The operator seems fine to me. But:
1) Why are there changes to pyinterpreter?
2) Do you have more context over the how this solves the problem? It seems like there should have been a design somewhere

Comment (laithsakka) [reply aten/src/ATen/native/native_functions.yaml:5517]: ok some context before this PR i will add that to the summary. 

we defined in this [PR](https://github.com/pytorch/pytorch/pull/157472) sym_is_contiguous to be the function computing contiguity for dynamic shapes. It returns a symbolic expression that represents contiguity and guaranteed not to throw a DDE.

when people call is_contiguous we do sym_is_contiguous().guard_bool()
when people call is_contiguous_or_false we do sym_is_contiguous().guard_or_false()
one issue not handled well was this path
```
c10::SymBool TensorImpl::sym_is_contiguous_custom(
    at::MemoryFormat memory_format) const {
  if (C10_UNLIKELY(matches_python_custom(SizesStridesPolicy::CustomStrides))) {
    return pyobj_slot_.load_pyobj_interpreter()->is_contiguous(
        this, memory_format);
  }

  return sym_is_contiguous_default(memory_format);
}
```
namely if we call sym_is_contiguous_custom but we have matches_python_custom(SizesStridesPolicy::CustomStrides) return true , then we used to call is_contiguous(this, memory_format); 

This used to go through the load_pyobj_interpreter and end up calling the python is_contiguous call which used implicit size oblivious reasoning.
once we removed that implicit size oblivious reasoning, the right thing we want is to call
return pyobj_slot_.load_pyobj_interpreter()->sym_is_contiguous(this, memory_format);
otherwise we would get DDE even if the caller is doing sym_is_contiguous.

so I had to define it for pyinterpreter, and then I had to override it for nested tensors.
make sense?

File: tools/autograd/gen_python_functions.py [hunk 4]
```diff
@@ -100,6 +100,7 @@
     "sym_size",
     "sym_stride",
     "sym_storage_offset",
+    "sym_is_contiguous",
```
Comment (aorenste) [inline tools/autograd/gen_python_functions.py]: please keep in alphabetical order

File: test/functorch/test_vmap_registrations.py [hunk 5]
```diff
@@ -209,6 +209,7 @@
     "aten::subtract_.Tensor",
     "aten::svd.U",
     "aten::sym_size.int",
+    "aten::sym_is_contiguous",
```
Comment (aorenste) [inline test/functorch/test_vmap_registrations.py:211]: please keep in alpha order

File: torch/_dynamo/convert_frame.py [hunk 6]
```diff
@@ -1527,7 +1527,6 @@ def __call__(
         frame_state: dict[str, Union[int, FrameStateSizeEntry]],
     ) -> ConvertFrameReturn:
         assert frame_state is not None
-
```
Comment (aorenste) [inline torch/_dynamo/convert_frame.py:1563]: unnecessary change

File: torch/_prims_common/__init__.py [hunk 7]
```diff
@@ -406,7 +406,7 @@ def is_channels_last_contiguous_or_false_3d(a: Tensor) -> bool:
 
 
 # similar to is_contiguous_for_memory_format but return false on data dependency.
-def contiguous_for_memory_format_or_false(  # type: ignore[return]
```
Comment (aorenste) [inline torch/_prims_common/__init__.py:409]: Do we need to worry about BC? I assume not since _prims_common starts with an underscore - but worth pointing out...

File: torch/csrc/PyInterpreter.cpp [hunk 8]
```diff
@@ -82,6 +82,8 @@ struct ConcretePyInterpreterVTable final
 
   bool is_contiguous(const c10::TensorImpl* self, at::MemoryFormat)
       const override;
+  c10::SymBool sym_is_contiguous(const c10::TensorImpl* self, at::MemoryFormat)
```
Comment (aorenste) [inline torch/csrc/PyInterpreter.cpp:85]: Maybe move down to be with the other `sym_` functions?

File: torch/csrc/PyInterpreter.cpp [hunk 9]
```diff
@@ -476,6 +478,33 @@ bool ConcretePyInterpreterVTable::is_contiguous(
...
+c10::SymBool ConcretePyInterpreterVTable::sym_is_contiguous(
+    const c10::TensorImpl* self,
+    at::MemoryFormat memory_format) const {
+  pybind11::gil_scoped_acquire gil;
+  at::impl::MaybeSetTLSOnEntryGuard guard;
+
+  py::object out;
+  out = torchDispatchFromTensorImpl(
+      self,
+      "is_contiguous",
```
Comment (aorenste) [inline torch/csrc/PyInterpreter.cpp]: It looks like this func_name is only used for printing errors, but maybe it should be "sym_is_contiguous"?

File: torch/masked/maskedtensor/_ops_refs.py [hunk 10]
```diff
@@ -285,7 +285,9 @@ def layout(func, *args, **kwargs):
     return _get_data(args[0]).layout
 
 
-@register_dispatch_func([torch.ops.aten.is_contiguous])
+@register_dispatch_func(
+    [torch.ops.aten.is_contiguous, torch.ops.aten.sym_is_contiguous]
+)
```
Comment (aorenste) [inline torch/masked/maskedtensor/_ops_refs.py:290]: I was wondering why this wasn't just an infinite recursive call - but I think what's happening is that when the dispatcher is calling this it temporarily removes this from the dispatch options so the recursive call will then call .default instead of .maskedtensor.

(nit: I'm sure it doesn't matter but below instead of reconstructing the args doing `return func(*args, **kwargs)` would be faster...)

File: torch/nested/_internal/nested_tensor.py [hunk 11]
```diff
@@ -239,9 +239,20 @@ def __repr__(self):  # type: ignore[override]
         grad_fn_str = (
             f", requires_grad={self.requires_grad}" if self.requires_grad else ""
         )
+
+        def is_contiguous_or_false():
```
Comment (aorenste) [inline torch/nested/_internal/nested_tensor.py]: nit: I would make this a `_is_contiguous_or_false` member. As it is it has to allocate a new lambda everytime it hits this...

File: torch/nested/_internal/ops.py [hunk 12]
```diff
@@ -516,6 +516,29 @@ def is_contiguous_general(func, *args, **kwargs):
...
+)
+def sym_is_contiguous_general(func, *args, **kwargs):
+    _, new_kwargs = normalize_function(  # type: ignore[misc]
+        func, args=args, kwargs=kwargs, normalize_to_only_use_kwargs=True
+    )
+    inp = new_kwargs.pop("input")
+
+    # If created from narrow() check for lengths
+    if inp.lengths() is not None:
+        return False
```
Comment (aorenste) [inline torch/nested/_internal/ops.py:530]: Maybe a quick comment why having lengths implies !contiguous

File: torch/_prims_common/__init__.py (hunk 1 above)
Comment (aorenste) [reply torch/_prims_common/__init__.py:271]: Does it make debugging harder? It early exits the false case and turns all the following guard_or_false/true below to just a simple "if", doesn't it?
(Same answer for is_channels_last_contiguous_2d)

File: aten/src/ATen/native/native_functions.yaml (hunk 3 above)
Comment (aorenste) [reply aten/src/ATen/native/native_functions.yaml:5517]: @laithsakka I think you made the summary better but maybe it could be better yet - maybe liberally steal from the doc I shared w/ you.
@zou3519 Does Laith's summary update address your concerns? This looks generally fine to me but I don't want to approve without checking for your input first.

File: torch/nested/_internal/ops.py [hunk 13]
```diff
@@ -516,6 +516,29 @@ def is_contiguous_general(func, *args, **kwargs):
 )(is_contiguous_general)
 
 
+@register_jagged_func(
```
Comment (ezyang) [inline torch/nested/_internal/ops.py:519]: I am not going to review this super carefully as NJT is resting

File: aten/src/ATen/native/native_functions.yaml (hunk 3 above)
Comment (ezyang) [reply aten/src/ATen/native/native_functions.yaml:5517]: The heuristic I used for this when I saw an earlier draft was that we had sym_size, and so, well, if this is just cargo culting, at least it's consistent lol

File: test/test_python_dispatch.py [hunk 14]
```diff
@@ -2002,6 +2002,8 @@ def __new__(cls, data, wrapper):
                 def __torch_dispatch__(cls, func, types, args, kwargs):
                     if func.overloadpacket == torch.ops.aten.is_contiguous:
                         return contiguous_data.is_contiguous()
+                    if func.overloadpacket == torch.ops.aten.sym_is_contiguous:
+                        return torch.ops.aten.sym_is_contiguous(contiguous_data)
```
Comment (ezyang) [inline test/test_python_dispatch.py:1962]: Hmm, I'm a little uncertain about this. If you are actually forcing all torch dispatch classes to add handling for `sym_is_contiguous` this is BC breaking. When there aren't symbolic integers anywhere in the Tensor, it would be better for BC if we still route to `is_contiguous`. I'm unsure how inconvenient this is, so I can be talked out of it, but needing to change it here is a smell.

File: torch/_prims_common/__init__.py (hunk 7 above)
Comment (ezyang) [reply torch/_prims_common/__init__.py:409]: This one I think is fine, pretty low risk. https://www.internalfb.com/code/search?q=repo%3Afbcode%20contiguous_for_memory_format_or_false

**Review by ezyang — APPROVED** [review]
Risky diff. Keep an eye out as it rolls out.

File: torch/_prims_common/__init__.py (hunk 1 above)
Comment (laithsakka) [reply torch/_prims_common/__init__.py:271]: I meant because it call c++ but i might just add do that for simplicity in a following PR

File: torch/nested/_internal/ops.py (hunk 12 above)
Comment (laithsakka) [reply torch/nested/_internal/ops.py:530]: this function is altered version of
def is_contiguous_general I am not sure what is the actual reasoning tbh.

File: test/functorch/test_vmap_registrations.py (hunk 5 above)
Comment (laithsakka) [reply test/functorch/test_vmap_registrations.py:211]: done

File: test/test_python_dispatch.py (hunk 14 above)
Comment (laithsakka) [reply test/test_python_dispatch.py:1962]: I see, I can do that . we can even restrict it to unhinted sizes/strides.

File: tools/autograd/gen_python_functions.py (hunk 4 above)
Comment (laithsakka) [reply tools/autograd/gen_python_functions.py]: done

File: torch/csrc/PyInterpreter.cpp (hunk 9 above)
Comment (laithsakka) [reply torch/csrc/PyInterpreter.cpp]: good catch lol

File: torch/nested/_internal/nested_tensor.py (hunk 11 above)
Comment (laithsakka) [reply torch/nested/_internal/nested_tensor.py]: done

**Comment by laithsakka** [comment]
published another identical PR
